modules/standard/Socket.md

```diff
@@ -0,0 +1,235 @@
+                  throwing `TimeoutError`. `(-1, 0)` implies indefinite blocking.
+```
+
+The purpose of family is important for address resolution and converting addr string to standard notation as well as validation. How the family can affect address resolution depends on [getaddrinfo](https://pubs.opengroup.org/onlinepubs/9699919799/functions/getaddrinfo.html) function.
+
+The `tcpConn` instance is basically a chapel `file` instance with added methods aligned to its nature as a socket. All the usual `file` methods are available on it:
+- `reader`
+  will return a reader channel for associated socket
+  ```python
+  proc tcpConn.writer() : channel
+  ```
+- `writer`
+  will return a writer channel for associated socket
+  ```python
+  proc tcpConn.writer() : channel
+  ```
+- `close` calling this on `tcpConn` will close the socket connection and associated `file` too.
+  ```python
+  proc tcpConn.close()
+  ```
+- `addr` this function will return a `ipAddr` object which will contain the `host` and `port` where socket is bound.
+
+## Creating a Listener
+
+`listen` function will create a `tcpListener` Instance which will take in following parameters
+
+**Method:**
+```python
+proc listen(in address:ipAddr, reuseAddr = true, backlog = 5) : tcpListener
+```
+
+**Parameters**
+```
+address: ipAddr - contains family info of socket to create and where to bind it.
+reuseaddr: Boolean - optional parameter, default value = true
+backlog: int - optional parameter, default value = 0
+```
+
+- The `backlog` argument defines the maximum length to which specifies the queue length for completely established sockets waiting to be accepted. If a connection request arrives when the queue is full, the client may receive an error with an indication of `ECONNREFUSED` or, if the underlying protocol supports retransmission, the request may be ignored so that a later reattempt at connection succeeds.
+- Calling in `listen` without `reuseaddr` can return an error that some other socket is already bound to the specified address. This can also be a result of re-running the program after only a short waiting period where the `bind` may find address as already in use.
+
+The `tcpListener` instance will have methods on it to call
+- `accept`: method will return a `tcpConn` instance of any new incoming connection otherwise will yield if no new connection is available at the instant. The communication can then proceed through the `tcpConn` instance returned
+  ```python
+  proc tcpListener.accept(in timeout:timeval = new timeval(-1,0)) : tcpConn
+  ```
+  `accept` will throw`TimeoutError` on not being able to accept any new connection
+  before `timeout`.
+- `close`,
+  calling close on `tcpListener` will close the socket server and it won't accept anymore new connections.
+  ```python
+  proc tcpListener.close()
+  ```
+- `family` returns the family of socket `AF_INET` or `AF_INET6`
```

Comment:
Shouldn't this return an instance of IPFamily instead?

Reply:
the function call will return a `c_int` via the c system call so is it possible to typecast it to `enum`?

Reply:
Yes we can cast integers to enums, e.g. `myint : myEnumType`. That cast can throw though. E.g.

```
enum myenum { a=1, b, c }
var i = 32;
var e = i:myenum;
writeln(e);
```

What might make sense in to have more than one getter function where one returns the OS view of the family and another returns the enum value. Anyway if we have the potential problem that the OS has more families than our enum value, we could include an `other` element in our enum.

Reply:
I mean, it's also possible to do a switch statement on the values returned and return the appropriate enum value for each without having to cast.

Reply:
The user can only set in the family using the `enum` itself, so if the OS has more families, it will be up to us to expand the options in `enum`; otherwise, the user can't use them; which also allows us to use `enum` casting as the family returned won't be out of the `enum` set.